### PR TITLE
(Closes #310) allow for current_scope to be None and extend tests

### LINKED ISSUE #310: Allow for matching of isolated code fragments (i.e. with no associated scoping region)
Testing PSyclone with master of fparser has revealed that the new symbol-table functionality causes problems - whenever fparser is used to parse just a fragment (e.g. by calling `match` directly) then there is no associated scoping region and `SYMBOL_TABLES.current_scope` (correctly) returns None. We just need to allow for this in the three places where `current_scope` is used.

### PR DESCRIPTION
**Pull request by arporter**

### CODE REVIEW

**Comment by arporter** [comment]
Ready for review from either @sergisiso or @rupertford.

**Review by rupertford — CHANGES_REQUESTED** [review]
Nice job. I've just requested a few more tests for the changes to Fortran2003.py

File: src/fparser/two/Fortran2003.py [hunk 1]
```diff
@@ -10281,7 +10283,9 @@ def match(string):
                 table.lookup(function_name)
                 # We found a matching name so refuse to match this intrinsic.
                 return None
-            except KeyError:
+            except (KeyError, AttributeError):
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/fparser/two/Fortran2003.py:10286]: Extended a test in `test_intrinsics.py`.

File: src/fparser/two/Fortran2003.py [hunk 2]
```diff
@@ -9308,12 +9309,13 @@ def match(string):
         result = Use_Stmt._match(string)
         if result:
             table = SYMBOL_TABLES.current_scope
-            only_list = None
-            # TODO #201 we currently ignore any symbol renaming here
-            if isinstance(result[4], Only_List):
-                names = walk(result[4], Name)
-                only_list = [name.string for name in names]
-            table.add_use_symbols(str(result[2]), only_list)
+            if table:
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/fparser/two/Fortran2003.py:9312]: Extended a test in `test_usestmt_r1109.py`

File: src/fparser/two/Fortran2003.py [hunk 3]
```diff
@@ -2684,7 +2684,7 @@ def match(string):
             # symbol table of the current scoping region.
             table = SYMBOL_TABLES.current_scope
 
-            if isinstance(result[0], Intrinsic_Type_Spec):
+            if table and isinstance(result[0], Intrinsic_Type_Spec):
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/fparser/two/Fortran2003.py:2687]: Done and moved associated test to its own file.

**Comment by arporter** [comment]
Ready for another look now.

File: src/fparser/two/tests/fortran2003/test_type_decl_stmt_r501.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
+# Copyright (c) 2022 Science and Technology Facilities Council.
```
Comment (rupertford) [inline src/fparser/two/tests/fortran2003/test_type_decl_stmt_r501.py:1]: Thanks for moving the tests into their own file. The only problem I have with this is that the tests have not been updated (apart from the symbol table part), they have been simply copied and I suspect they will need to be improved/extended e.g. what about `match2`. Could you add a new issue to do this and reference it here. You could also reference #259 from here too.

**Review by rupertford — CHANGES_REQUESTED** [review]
Just the one issue.

File: src/fparser/two/tests/fortran2003/test_type_decl_stmt_r501.py (hunk 4 above)
Comment (arporter) [reply src/fparser/two/tests/fortran2003/test_type_decl_stmt_r501.py:1]: Done.

**Review by rupertford — APPROVED** [review]
All looks good. Approving.

**Pull request merged**